Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+* The "muxer" (`dotnet.exe`)
+* The `apphost` (`.exe` which is part of the app)
+* The `comhost` (`.dll` which is part of the app and acts as COM server)
+* The `ijwhost` (`.dll` consumed via `.lib` used by IJW assemblies)
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native apps to easily host managed components and to easily locate `hostfxr` for more advanced scenarios.
+
+At the same time add the ability to pass additional runtime properties when starting the runtime through the hosting entry points (starting app, loading component). This can be used by the native app to:
+* Register startup hook without modifying environment variables (which are inherited by child processes)
+* Introduce new runtime knobs which are only available for native hosts without the need to update the hosting APIs every time.
+
+This new ability would be added to both `hostfxr` and to the `nethost`.
+
+
+## New host binary for component hosting
+Add new library `nethost` which will act as the easy to use host for loading managed components.
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dylib`). For ease of use there would be a header file for C++ apps as well as `.lib`/`.a` for easy linking.
+Apps using the component hosting functionality would ship this library as part of the app. So similarly to `apphost`/`comhost`/`ijwhost` the `nethost` will be shipped as part of .NET Core SDK to be included in customer's projects.
+
+This library would expose two APIs
+
+### Load managed component and get a function pointer
+``` C++
+int nethost_load_assembly_method(
+        const char * assembly_path,
+        const char * type_name,
+        const char * method_name,
+        const void * reserved,
+        void ** delegate);
+)
+```
+This API will
+* Locate the assembly using the `assembly_path` and its `.runtimeconfig.json` and determine the frameworks it requires to run. (Note that only framework dependent components will be supported for now).
+* If the process doesn't have CoreCLR loaded (more specifically `hostpolicy` library)
+  * Using the `.runtimeconfig.json` resolve required frameworks (just like if running an application) and load the runtime.
+* Else the CoreCLR is already loaded, in that case validate that required frameworks for the component can be satisfied by the runtime.
+  * If the required frameworks are not already present, fail. No support to load additional frameworks for now.
+* Call into the runtime (`System.Private.CoreLib` specifically)
+  * Create a new isolated `AssemblyLoadContext` (possibly reusing for the same components)
```

Comment:
Is this new? Why is that a feature of this API, what if we just want it loaded into the default context?

Reply:
The entire "ability to load managed components into native processes" is a new feature (although quite frequently requested).

This isolation behavior will be the same for `comhost`, `nethost` and `ijwhost` which all implement that same underlying principle - loading a managed component into "unknown" process.

The full isolation is needed to make sure that one can load any combination of managed components into any process. There's the added problem of frameworks which have to match, but that's for a longer discussion (let's just say the recommendation and hopefully the defualt will be to use the new `RollForward=LatestMajor` for the components to maximaze chance of successful framework resolution).

Also at least some isolation is needed to avoid failure in case two components have similar but different dependencies (two versions of JSON for example).

There are basically 2 main scenarios:

#### Native C++ app (or any other non-.NET app) loading managed components
For this case the plan is to populate the default ALC with only the frameworks. The reason is because default ALC is special (acts as a fallback for all other ALCs), and there's no good way to pick one of the components as "the main", so none will get that priviledge.

#### App which already has .NET Core in it
This could be a .NET Core app which calls into native which in turn tries to load managed component or it might be a more complex case of some special native host.
The default ALC in this case is basically unpredictable - it can contain almost anything. That said we will know what frameworks it will contain. So just like above, the new loaded component will go into fully isolated ALC to avoid collisions with anything in the default ALC, reusing only the framework from the default ALC.

Given the scenarios above - what kind of scenario do you think would want to load the managed component into the default ALC (or rather would need to do that).

I just want to reiterate: This feature is for loading components (think `classlib`), NOT for loading apps.